File: python/scripts/smoke_test.py
#!/usr/bin/env python3
"""
Smoke Test - Verify Python SDK works end-to-end with real APIs

Run: uv run python scripts/smoke_test.py

This script tests all major features and outputs results.
After running, manually verify traces appear in the Raindrop dashboard.
"""

import os
import sys
import time
from dataclasses import dataclass

# Check env vars before importing SDK
RAINDROP_API_KEY = os.environ.get("RAINDROP_API_KEY")
OPENAI_API_KEY = os.environ.get("OPENAI_API_KEY")

if not RAINDROP_API_KEY or not OPENAI_API_KEY:
    print("Missing required environment variables:")
    print(f"  RAINDROP_API_KEY: {'✓' if RAINDROP_API_KEY else '✗'}")
    print(f"  OPENAI_API_KEY: {'✓' if OPENAI_API_KEY else '✗'}")
    sys.exit(1)

from openai import OpenAI

from rd_mini import Raindrop


# ============================================
# Helpers
# ============================================


@dataclass
class TestResult:
    test: str
    status: str  # PASS or FAIL
    trace_id: str | None = None
    error: str | None = None


results: list[TestResult] = []


def log(message: str) -> None:
    print(f"\n{'=' * 60}\n{message}\n{'=' * 60}")


def passed(test: str, trace_id: str | None = None) -> None:
    results.append(TestResult(test, "PASS", trace_id))
    print(f"  ✓ {test}" + (f" ({trace_id})" if trace_id else ""))


def failed(test: str, error: str) -> None:
    results.append(TestResult(test, "FAIL", error=error))
    print(f"  ✗ {test}: {error}")


# ============================================
# Tests
# ============================================


def main() -> None:
    print("\n🧪 RAINDROP PYTHON SDK SMOKE TEST\n")
    print(f"API Key: {RAINDROP_API_KEY[:8]}...")
    print(f"Timestamp: {time.strftime('%Y-%m-%dT%H:%M:%S')}")
    print("")

    # Initialize
    raindrop = Raindrop(
        api_key=RAINDROP_API_KEY,
        debug=True,
    )

    openai_client = raindrop.wrap(OpenAI(api_key=OPENAI_API_KEY))

    # ------------------------------------------
    log("TEST 1: Basic Chat Completion")
    # ------------------------------------------
    try:
        response = openai_client.chat.completions.create(
            model="gpt-4o-mini",
            messages=[{"role": "user", "content": 'Say "Python smoke test" in exactly 3 words.'}],
        )

        if hasattr(response, "_trace_id") and response.choices[0].message.content:
            passed("Non-streaming chat completion", response._trace_id)
            print(f'    Response: "{response.choices[0].message.content}"')
            print(f"    Tokens: {response.usage.total_tokens if response.usage else 'N/A'}")
        else:
            failed("Non-streaming chat completion", "Missing _trace_id or content")
    except Exception as e:
        failed("Non-streaming chat completion", str(e))

    # ------------------------------------------
    log("TEST 2: Streaming Chat Completion")
    # ------------------------------------------
    try:
        stream = openai_client.chat.completions.create(
            model="gpt-4o-mini",
            messages=[{"role": "user", "content": "Count from 1 to 5, one number per line."}],
            stream=True,
        )

        if hasattr(stream, "_trace_id"):
            passed("Stream has _trace_id immediately", stream._trace_id)
        else:
            failed("Stream has _trace_id immediately", "Missing _trace_id on stream")

        content = ""
        for chunk in stream:
            delta = chunk.choices[0].delta if chunk.choices else None
            if delta and delta.content:
                content += delta.content

        if "1" in content and "5" in content:
            passed("Stream content received")
            print(f'    Content: "{content[:50]}..."')
        else:
            failed("Stream content received", "Missing expected numbers")
    except Exception as e:
        failed("Streaming chat completion", str(e))

    # ------------------------------------------
    log("TEST 3: Tool Calls")
    # ------------------------------------------
    try:
        response = openai_client.chat.completions.create(
            model="gpt-4o-mini",
            messages=[{"role": "user", "content": "What is the weather in San Francisco?"}],
            tools=[
                {
                    "type": "function",
                    "function": {
                        "name": "get_weather",
                        "description": "Get weather for a location",
                        "parameters": {
                            "type": "object",
                            "properties": {"location": {"type": "string"}},
                            "required": ["location"],
                        },
                    },
                }
            ],
            tool_choice="auto",
        )

        tool_calls = response.choices[0].message.tool_calls
        if tool_calls and tool_calls[0].function.name == "get_weather":
            passed("Tool call extracted", response._trace_id)
            print(f"    Tool: {tool_calls[0].function.name}")
            print(f"    Args: {tool_calls[0].function.arguments}")
        else:
            failed("Tool call extracted", "No tool call in response")
    except Exception as e:
        failed("Tool calls", str(e))

    # ------------------------------------------
    log("TEST 4: User Identification")
    # ------------------------------------------
    try:
        test_user_id = f"py_smoke_test_{int(time.time())}"
        raindrop.identify(
            test_user_id,
            {
                "name": "Python Smoke Test User",
                "email": "pysmoke@test.com",
                "plan": "enterprise",
            },
        )
        passed("User identified")
        print(f"    User ID: {test_user_id}")
    except Exception as e:
        failed("User identification", str(e))

    # ------------------------------------------
    log("TEST 5: Conversation Threading")
    # ------------------------------------------
    try:
        conversation_id = f"py_smoke_convo_{int(time.time())}"

        # Note: Python SDK uses raindrop= kwarg, not extra_body
        msg1 = openai_client.chat.completions.create(
            model="gpt-4o-mini",
            messages=[{"role": "user", "content": 'Remember the word "giraffe".'}],
            raindrop={"conversation_id": conversation_id},
        )

        msg2 = openai_client.chat.completions.create(
            model="gpt-4o-mini",
            messages=[
                {"role": "user", "content": 'Remember the word "giraffe".'},
                {"role": "assistant", "content": msg1.choices[0].message.content or ""},
                {"role": "user", "content": "What word did I ask you to remember?"},
            ],
            raindrop={"conversation_id": conversation_id},
        )

        passed("Conversation threading")
        print(f"    Conversation ID: {conversation_id}")
        print(f"    Message 1: {msg1._trace_id}")
        print(f"    Message 2: {msg2._trace_id}")
    except Exception as e:
        failed("Conversation threading", str(e))

    # ------------------------------------------
    log("TEST 6: Feedback / Signals")
    # ------------------------------------------
    try:
        response = openai_client.chat.completions.create(
            model="gpt-4o-mini",
            messages=[{"role": "user", "content": "Tell me a one-liner joke."}],
        )

        raindrop.feedback(
            response._trace_id,
            {
                "type": "thumbs_up",
                "comment": "Python smoke test - feedback works!",
            },
        )

        passed("Feedback sent", response._trace_id)
        print("    Type: thumbs_up")
        print(f"    Linked to: {response._trace_id}")
    except Exception as e:
        failed("Feedback", str(e))

    # ------------------------------------------
    log("TEST 7: Interaction Context Manager")
    # ------------------------------------------
    try:
        with raindrop.interaction(
            event="py_smoke_test_rag",
            input="Python test query",
        ) as ctx:
            # Simulate a tool call
            @raindrop.tool("search_docs")
            def search_docs(query: str) -> list[dict]:
                return [{"title": "Doc 1", "content": "Python content"}]

            search_results = search_docs("test query")

            # AI call within interaction
            response = openai_client.chat.completions.create(
                model="gpt-4o-mini",
                messages=[{"role": "user", "content": f"Summarize: {search_results}"}],
            )

            ctx.output = response.choices[0].message.content or ""
            interaction_id = ctx.interaction_id

        passed("Interaction with spans")
        print(f"    Interaction ID: {interaction_id}")
        print("    Event: py_smoke_test_rag")
    except Exception as e:
        failed("Interaction with spans", str(e))

    # ------------------------------------------
    log("TEST 8: Begin/Finish Pattern")
    # ------------------------------------------
    try:
        interaction = raindrop.begin(
            event="py_smoke_test_manual",
            input="Python manual interaction test",
        )

        # Simulate some work
        time.sleep(0.1)

        interaction.output = "Python manual interaction completed"
        interaction.finish()

        passed("Begin/finish pattern")
        print(f"    Interaction ID: {interaction.id}")
    except Exception as e:
        failed("Begin/finish pattern", str(e))

    # ------------------------------------------
    log("TEST 9: @workflow Decorator")
    # ------------------------------------------
    try:

        @raindrop.workflow("py_smoke_workflow")
        def my_workflow(message: str) -> str:
            response = openai_client.chat.completions.create(
                model="gpt-4o-mini",
                messages=[{"role": "user", "content": message}],
            )
            return response.choices[0].message.content or ""

        result = my_workflow("What is 2+2?")
        passed("Workflow decorator")
        print(f"    Result: {result[:50]}...")
    except Exception as e:
        failed("Workflow decorator", str(e))

    # ------------------------------------------
    log("TEST 10: Error Handling")
    # ------------------------------------------
    try:
        openai_client.chat.completions.create(
            model="not-a-real-model",
            messages=[{"role": "user", "content": "test"}],
        )
        failed("Error handling", "Expected error to be thrown")
    except Exception as e:
        last_trace_id = raindrop.get_last_trace_id()
        if last_trace_id:
            passed("Error traced", last_trace_id)
            print(f"    Error: {str(e)[:60]}...")
        else:
            failed("Error handling", "Error not traced")

    # ------------------------------------------
    log("TEST 11: Flush & Close")
    # ------------------------------------------
    try:
        raindrop.flush()
        passed("Flush completed")

        raindrop.close()
        passed("Close completed")
    except Exception as e:
        failed("Flush/close", str(e))

    # ------------------------------------------
    # Summary
    # ------------------------------------------
    log("SUMMARY")

    passed_count = sum(1 for r in results if r.status == "PASS")
    failed_count = sum(1 for r in results if r.status == "FAIL")

    print(f"\n  Passed: {passed_count}")
    print(f"  Failed: {failed_count}")
    print(f"  Total:  {len(results)}")

    if failed_count > 0:
        print("\n  Failed tests:")
        for r in results:
            if r.status == "FAIL":
                print(f"    - {r.test}: {r.error}")

    print("\n" + "=" * 60)
    print("NEXT STEPS:")
    print("=" * 60)
    print("\n1. Open Raindrop dashboard")
    print("2. Verify these traces appear within 30 seconds:")
    for r in results:
        if r.trace_id:
            print(f"   - {r.trace_id} ({r.test})")
    print("\n3. Check:")
    print("   - [ ] Input/output content visible")
    print("   - [ ] Token counts displayed")
    print("   - [ ] User identified correctly")
    print("   - [ ] Conversation traces grouped")
    print("   - [ ] Feedback appears on trace")
    print("   - [ ] Spans visible in interaction")
    print("")

    sys.exit(1 if failed_count > 0 else 0)


if __name__ == "__main__":
    main()

File: ts/scripts/smoke-test.ts
#!/usr/bin/env bun
/**
 * Smoke Test - Verify SDK works end-to-end with real APIs
 *
 * Run: bun run scripts/smoke-test.ts
 *
 * This script tests all major features and outputs results.
 * After running, manually verify traces appear in the Raindrop dashboard.
 */

import { Raindrop } from '../src/index.js';
import OpenAI from 'openai';

// ============================================
// Config
// ============================================

const RAINDROP_API_KEY = process.env.RAINDROP_API_KEY;
const OPENAI_API_KEY = process.env.OPENAI_API_KEY;

if (!RAINDROP_API_KEY || !OPENAI_API_KEY) {
  console.error('Missing required environment variables:');
  console.error('  RAINDROP_API_KEY:', RAINDROP_API_KEY ? '✓' : '✗');
  console.error('  OPENAI_API_KEY:', OPENAI_API_KEY ? '✓' : '✗');
  process.exit(1);
}

// ============================================
// Helpers
// ============================================

const results: Array<{ test: string; status: 'PASS' | 'FAIL'; traceId?: string; error?: string }> = [];

function log(message: string) {
  console.log(`\n${'='.repeat(60)}\n${message}\n${'='.repeat(60)}`);
}

function pass(test: string, traceId?: string) {
  results.push({ test, status: 'PASS', traceId });
  console.log(`  ✓ ${test}${traceId ? ` (${traceId})` : ''}`);
}

function fail(test: string, error: string) {
  results.push({ test, status: 'FAIL', error });
  console.log(`  ✗ ${test}: ${error}`);
}

// ============================================
// Tests
// ============================================

async function main() {
  console.log('\n🧪 RAINDROP SDK SMOKE TEST\n');
  console.log('API Key:', RAINDROP_API_KEY?.slice(0, 8) + '...');
  console.log('Timestamp:', new Date().toISOString());
  console.log('');

  // Initialize
  const raindrop = new Raindrop({
    apiKey: RAINDROP_API_KEY,
    debug: true,  // Show all sends
  });

  const openai = raindrop.wrap(new OpenAI({ apiKey: OPENAI_API_KEY }));

  try {
    // ------------------------------------------
    log('TEST 1: Basic Chat Completion');
    // ------------------------------------------
    try {
      const response = await openai.chat.completions.create({
        model: 'gpt-4o-mini',
        messages: [{ role: 'user', content: 'Say "smoke test successful" in exactly 3 words.' }],
      });

      if (response._traceId && response.choices[0]?.message?.content) {
        pass('Non-streaming chat completion', response._traceId);
        console.log(`    Response: "${response.choices[0].message.content}"`);
        console.log(`    Tokens: ${response.usage?.total_tokens || 'N/A'}`);
      } else {
        fail('Non-streaming chat completion', 'Missing traceId or content');
      }
    } catch (e) {
      fail('Non-streaming chat completion', String(e));
    }

    // ------------------------------------------
    log('TEST 2: Streaming Chat Completion');
    // ------------------------------------------
    try {
      const stream = await openai.chat.completions.create({
        model: 'gpt-4o-mini',
        messages: [{ role: 'user', content: 'Count from 1 to 5, one number per line.' }],
        stream: true,
      });

      if (stream._traceId) {
        pass('Stream has traceId immediately', stream._traceId);
      } else {
        fail('Stream has traceId immediately', 'Missing _traceId on stream');
      }

      let content = '';
      for await (const chunk of stream) {
        content += chunk.choices[0]?.delta?.content || '';
      }

      if (content.includes('1') && content.includes('5')) {
        pass('Stream content received');
        console.log(`    Content: "${content.slice(0, 50)}..."`);
      } else {
        fail('Stream content received', 'Missing expected numbers');
      }
    } catch (e) {
      fail('Streaming chat completion', String(e));
    }

    // ------------------------------------------
    log('TEST 3: Tool Calls');
    // ------------------------------------------
    try {
      const response = await openai.chat.completions.create({
        model: 'gpt-4o-mini',
        messages: [{ role: 'user', content: 'What is the weather in San Francisco?' }],
        tools: [{
          type: 'function',
          function: {
            name: 'get_weather',
            description: 'Get weather for a location',
            parameters: {
              type: 'object',
              properties: { location: { type: 'string' } },
              required: ['location'],
            },
          },
        }],
        tool_choice: 'auto',
      });

      const toolCall = response.choices[0]?.message?.tool_calls?.[0];
      if (toolCall?.function?.name === 'get_weather') {
        pass('Tool call extracted', response._traceId);
        console.log(`    Tool: ${toolCall.function.name}`);
        console.log(`    Args: ${toolCall.function.arguments}`);
      } else {
        fail('Tool call extracted', 'No tool call in response');
      }
    } catch (e) {
      fail('Tool calls', String(e));
    }

    // ------------------------------------------
    log('TEST 4: User Identification');
    // ------------------------------------------
    try {
      const testUserId = `smoke_test_${Date.now()}`;
      raindrop.identify(testUserId, {
        name: 'Smoke Test User',
        email: 'smoke@test.com',
        plan: 'enterprise',
      });
      pass('User identified');
      console.log(`    User ID: ${testUserId}`);
    } catch (e) {
      fail('User identification', String(e));
    }

    // ------------------------------------------
    log('TEST 5: Conversation Threading');
    // ------------------------------------------
    try {
      const conversationId = `smoke_convo_${Date.now()}`;

      const msg1 = await openai.chat.completions.create(
        {
          model: 'gpt-4o-mini',
          messages: [{ role: 'user', content: 'Remember the word "elephant".' }],
        },
        { raindrop: { conversationId } }
      );

      const msg2 = await openai.chat.completions.create(
        {
          model: 'gpt-4o-mini',
          messages: [
            { role: 'user', content: 'Remember the word "elephant".' },
            { role: 'assistant', content: msg1.choices[0].message.content || '' },
            { role: 'user', content: 'What word did I ask you to remember?' },
          ],
        },
        { raindrop: { conversationId } }
      );

      pass('Conversation threading');
      console.log(`    Conversation ID: ${conversationId}`);
      console.log(`    Message 1: ${msg1._traceId}`);
      console.log(`    Message 2: ${msg2._traceId}`);
    } catch (e) {
      fail('Conversation threading', String(e));
    }

    // ------------------------------------------
    log('TEST 6: Feedback / Signals');
    // ------------------------------------------
    try {
      const response = await openai.chat.completions.create({
        model: 'gpt-4o-mini',
        messages: [{ role: 'user', content: 'Tell me a one-liner joke.' }],
      });

      raindrop.feedback(response._traceId, {
        type: 'thumbs_up',
        comment: 'Smoke test - feedback works!',
      });

      pass('Feedback sent', response._traceId);
      console.log(`    Type: thumbs_up`);
      console.log(`    Linked to: ${response._traceId}`);
    } catch (e) {
      fail('Feedback', String(e));
    }

    // ------------------------------------------
    log('TEST 7: Interaction with Spans');
    // ------------------------------------------
    try {
      const interactionId = await raindrop.withInteraction(
        {
          event: 'smoke_test_rag',
          input: 'Test query for smoke test',
        },
        async (ctx) => {
          // Simulate a tool call
          const searchResults = await raindrop.wrapTool(
            'search_docs',
            async (query: string) => {
              // Fake search
              return [{ title: 'Doc 1', content: 'Relevant content' }];
            }
          )('test query');

          // AI call within interaction
          const response = await openai.chat.completions.create({
            model: 'gpt-4o-mini',
            messages: [{ role: 'user', content: 'Summarize: ' + JSON.stringify(searchResults) }],
          });

          ctx.output = response.choices[0].message.content || '';

          return ctx.interactionId;
        }
      );

      pass('Interaction with spans');
      console.log(`    Interaction ID: ${interactionId}`);
      console.log(`    Event: smoke_test_rag`);
    } catch (e) {
      fail('Interaction with spans', String(e));
    }

    // ------------------------------------------
    log('TEST 8: Begin/Finish Pattern');
    // ------------------------------------------
    try {
      const interaction = raindrop.begin({
        event: 'smoke_test_manual',
        input: 'Manual interaction test',
      });

      // Simulate some work
      await new Promise((resolve) => setTimeout(resolve, 100));

      interaction.output = 'Manual interaction completed';
      interaction.finish();

      pass('Begin/finish pattern');
      console.log(`    Interaction ID: ${interaction.id}`);
    } catch (e) {
      fail('Begin/finish pattern', String(e));
    }

    // ------------------------------------------
    log('TEST 9: Error Handling');
    // ------------------------------------------
    try {
      await openai.chat.completions.create({
        model: 'not-a-real-model',
        messages: [{ role: 'user', content: 'test' }],
      });
      fail('Error handling', 'Expected error to be thrown');
    } catch (e) {
      const lastTraceId = raindrop.getLastTraceId();
      if (lastTraceId) {
        pass('Error traced', lastTraceId);
        console.log(`    Error: ${String(e).slice(0, 60)}...`);
      } else {
        fail('Error handling', 'Error not traced');
      }
    }

    // ------------------------------------------
    log('TEST 10: Flush & Close');
    // ------------------------------------------
    try {
      await raindrop.flush();
      pass('Flush completed');

      await raindrop.close();
      pass('Close completed');
    } catch (e) {
      fail('Flush/close', String(e));
    }

  } catch (e) {
    console.error('\n\nFATAL ERROR:', e);
    process.exit(1);
  }

  // ------------------------------------------
  // Summary
  // ------------------------------------------
  log('SUMMARY');

  const passed = results.filter((r) => r.status === 'PASS').length;
  const failed = results.filter((r) => r.status === 'FAIL').length;

  console.log(`\n  Passed: ${passed}`);
  console.log(`  Failed: ${failed}`);
  console.log(`  Total:  ${results.length}`);

  if (failed > 0) {
    console.log('\n  Failed tests:');
    results.filter((r) => r.status === 'FAIL').forEach((r) => {
      console.log(`    - ${r.test}: ${r.error}`);
    });
  }

  console.log('\n' + '='.repeat(60));
  console.log('NEXT STEPS:');
  console.log('='.repeat(60));
  console.log('\n1. Open Raindrop dashboard');
  console.log('2. Verify these traces appear within 30 seconds:');
  results.filter((r) => r.traceId).forEach((r) => {
    console.log(`   - ${r.traceId} (${r.test})`);
  });
  console.log('\n3. Check:');
  console.log('   - [ ] Input/output content visible');
  console.log('   - [ ] Token counts displayed');
  console.log('   - [ ] User identified correctly');
  console.log('   - [ ] Conversation traces grouped');
  console.log('   - [ ] Feedback appears on trace');
  console.log('   - [ ] Spans visible in interaction');
  console.log('');

  process.exit(failed > 0 ? 1 : 0);
}

main();
